src/Cafe/OS/libs/nsyshid/Dimensions.cpp

```diff
@@ -0,0 +1,1076 @@
+										  uint8((dataOne >> 16) & 0xFF), uint8((dataOne >> 24) & 0xFF),
+										  uint8(dataTwo & 0xFF), uint8((dataTwo >> 8) & 0xFF),
+										  uint8((dataTwo >> 16) & 0xFF), uint8((dataTwo >> 24) & 0xFF)};
+		return encrypted;
+	}
+
+	std::array<uint8, 16> DimensionsUSB::GenerateFigureKey(const std::array<uint8, 0x2D * 0x04>& buf)
+	{
+		std::array<uint8, 7> uid = {buf[0], buf[1], buf[2], buf[4], buf[5], buf[6], buf[7]};
+
+		uint32 scrambleA = Scramble(uid, 3);
+		uint32 scrambleB = Scramble(uid, 4);
+		uint32 scrambleC = Scramble(uid, 5);
+		uint32 scrambleD = Scramble(uid, 6);
+
+		return {uint8((scrambleA >> 24) & 0xFF), uint8((scrambleA >> 16) & 0xFF),
+				uint8((scrambleA >> 8) & 0xFF), uint8(scrambleA & 0xFF),
+				uint8((scrambleB >> 24) & 0xFF), uint8((scrambleB >> 16) & 0xFF),
+				uint8((scrambleB >> 8) & 0xFF), uint8(scrambleB & 0xFF),
+				uint8((scrambleC >> 24) & 0xFF), uint8((scrambleC >> 16) & 0xFF),
+				uint8((scrambleC >> 8) & 0xFF), uint8(scrambleC & 0xFF),
+				uint8((scrambleD >> 24) & 0xFF), uint8((scrambleD >> 16) & 0xFF),
+				uint8((scrambleD >> 8) & 0xFF), uint8(scrambleD & 0xFF)};
+	}
+
+	uint32 DimensionsUSB::Scramble(const std::array<uint8, 7>& uid, uint8 count)
+	{
+		std::vector<uint8> toScramble;
+		toScramble.reserve(uid.size() + CHAR_CONSTANT.size());
+		for (uint8 x : uid)
+		{
+			toScramble.push_back(x);
+		}
+		for (uint8 c : CHAR_CONSTANT)
+		{
+			toScramble.push_back(c);
+		}
+		toScramble[(count * 4) - 1] = 0xaa;
+
+		std::array<uint8, 4> randomized = DimensionsRandomize(toScramble, count);
+
+		return uint32(randomized[0]) << 24 | uint32(randomized[1]) << 16 | uint32(randomized[2]) << 8 | uint32(randomized[3]);
+	}
+
+	std::array<uint8, 4> DimensionsUSB::PWDGenerate(const std::array<uint8, 0x2D * 0x04>& buf)
+	{
+		std::array<uint8, 7> uid = {buf[0], buf[1], buf[2], buf[4], buf[5], buf[6], buf[7]};
+
+		std::vector<uint8> pwdCalc = {PWD_CONSTANT.begin(), PWD_CONSTANT.end() - 1};
+		for (uint8 i = 0; i < uid.size(); i++)
+		{
+			pwdCalc.insert(pwdCalc.begin() + i, uid[i]);
+		}
+
+		return DimensionsRandomize(pwdCalc, 8);
+	}
+
+	std::array<uint8, 4> DimensionsUSB::DimensionsRandomize(const std::vector<uint8> key, uint8 count)
+	{
+		uint32 scrambled = 0;
+		for (uint8 i = 0; i < count; i++)
+		{
+			const uint32 v4 = std::rotr(scrambled, 25);
+			const uint32 v5 = std::rotr(scrambled, 10);
+			const uint32 b = uint32(key[(i * 4) + 3]) << 24 | uint32(key[(i * 4) + 2]) << 16 | uint32(key[(i * 4) + 1]) << 8 | uint32(key[(i * 4)]);
+			scrambled = b + v4 + v5 - scrambled;
+		}
+		return {uint8(scrambled & 0xFF), uint8(scrambled >> 8 & 0xFF), uint8(scrambled >> 16 & 0xFF), uint8(scrambled >> 24 & 0xFF)};
+	}
+
+	uint32 DimensionsUSB::GetFigureId(const std::array<uint8, 0x2D * 0x04>& buf)
+	{
+		const std::array<uint8, 16> figureKey = GenerateFigureKey(buf);
+
+		const std::array<uint8, 8> decrypted = Decrypt(&buf[36 * 4], figureKey);
```

Comment:
If you do change the argument type to std::span, ``&buf[36 * 4]`` would need to be replaced with something like ``std::span<const uint8, 8>{buf.begin() + 36 * 4, 8}`` or shorter but less C++-like ``(const uint8(&)[8])buf[36 * 4]``, both of which are fairly lengthy for a function argument.
Considering that in this case you access the same area of the buffer later in the function you could define the span as a local variable that you use every time you want to access something at 36 * 4. That way the argument for Decrypt can be short, and you repeat that index less often so the code is less prone to typo's. (not that it matters if you wrote it correctly, but "Do not Repeat Yourself" is good practice regardless)

Reply:
I have had a go at using span across the class - let me know if my implementations look correct, haven't got a whole lot of C++ experience besides the occasional open source contribution, so learning new things is always a bonus here (especially best practices for C++ 20)

Reply:
Looks good except for what I pointed out below in a different comment